tempesta_fw/http2/huffman.c

```diff
@@ -0,0 +1,641 @@
+		shift = state[i].shift;
+		if (shift >= 0) {
+			if (shift > current + HT_NBITS) {
+				break;
+			}
+			* dst++ = (char) state[i].offset;
+			current -= shift;
+			state = HTDecode;
+		}
+		else {
+			/*
+			 * Last full prefix also processed here, to allow
+			 * EOS padding detection. Condition here equivalent to
+			 * the "-shift >= current + HT_NBITS", but working faster:
+			 */
+			if (Opt_Likely(shift <= -HT_NBITS - current)) {
+				break;
+			}
+			return state[i].offset == 0 ? HTTP2Error_Huffman_UnexpectedEOS :
+						      HTTP2Error_Huffman_InvalidCode;
+		}
+	}
+	if (Opt_Likely(state == HTDecode &&
+	       (i ^ (HT_EOS_HIGH >> 1)) < (1U << -current)))
+	{
+		return 0;
+	}
+	else {
+		return HTTP2Error_Huffman_CodeTooShort;
+	}
+}
+
+static fast
+http2_huffman_decode_tail_s (ufast			c,
+			     char	   * __restrict dst,
+			     fast			current,
+			     const HTState * __restrict state)
+{
+	int16 offset;
+	fast shift;
+	ufast i;
+	if (Opt_Unlikely(current == -HT_MBITS)) {
+		return HTTP2Error_Huffman_CodeTooShort;
+	}
+	i = (c << -current) & HT_MMASK;
+	shift = state[i].shift;
+	offset = state[i].offset;
+	if (Opt_Likely(shift >= 0)) {
+		if (Opt_Unlikely(shift > current + HT_NBITS)) {
+			return HTTP2Error_Huffman_CodeTooShort;
+		}
+		* dst++ = (char) offset;
+		current -= shift;
+		return http2_huffman_decode_tail(c, dst, current, HTDecode);
+	}
+	else {
+		/*
+		 * Condition here equivalent to the "-shift > current + HT_NBITS",
+		 * but working faster:
+		 */
+		if (Opt_Unlikely(shift < -HT_NBITS - current)) {
+			return HTTP2Error_Huffman_CodeTooShort;
+		}
+		return offset == 0 ? HTTP2Error_Huffman_UnexpectedEOS :
+				     HTTP2Error_Huffman_InvalidCode;
+	}
+}
+
+fast
+http2_huffman_decode (const char * __restrict source,
+			    char * __restrict dst,
+			    uwide	      n)
+{
+	if (n) {
+		const uchar * __restrict src = (const uchar *) source;
+		ufast c = * src++;
+		fast current = 8 - HT_NBITS;
+		int16 offset;
+		n--;
+		for (;;) {
+			const HTState * __restrict state;
+Root:			state = HTDecode;
+			for (;;) {
+				fast shift;
+				ufast i;
+				if (current <= 0) {
+					if (Opt_Likely(n)) {
+						current += 8;
+						c = Bit_Join8(c, * src++);
+						n--;
+					}
+					else {
+					     /* Last full prefix also processed here */
+					     /* (see current <= 0 above): */
+						return http2_huffman_decode_tail(
+							c, dst, current, state
+						);
+					}
+				}
+				i = (c >> current) & HT_NMASK;
+				shift = state[i].shift;
+				offset = state[i].offset;
+				if (shift >= 0) {
+					* dst++ = (char) offset;
+					current -= shift;
+					goto Root;
+				}
+				else {
+					current += shift;
+				#ifdef HT_BALANCED_TREE
+					if (Opt_Likely(offset)) {
+				#else
+					if (Opt_Likely(offset > 0)) {
+				#endif
+						state = HTDecode + offset;
+						if (offset >= HT_SMALL) {
+							break;
+						}
+					}
+					else {
+						goto End;
+					}
+				}
+			}
+			current += HT_NBITS - HT_MBITS;
+			{
```

Comment:
The values ​​of the variables "I" and "shift" in the first block are not related to the values ​​of the variables "I" and "shift" in the second block. If we declare them externally, then we force the compiler to allocate the same registers for two different sets of variables. Probably, SSA-analysis in the latest versions of GCC will be able to split the false dependency on the data by using analysis of the post-dominantors tree, since it will see that these variables are re-assigned anew outside the loop. And it himself will create two new variables for the second block in the new registers. But I tried to write code in such manner that it does not much depend on the features of optimization algorithms in the compiler. One of the main general rules is to minimize the scope of all independent variables.